[docs] Add documentation for hybrid commands (#8244)
Rapptz/discord.py: docs/ext/commands/commands.rst

    Be careful on how you write your global checks, as it could also lock you out of your own bot.

.. need a note on global check once here I think

Hybrid Commands
---------------

.. versionadded:: 2.0

> :class:`.commands.HybridCommand` is a command that can be invoked as both a text and a slash command.

Reviewer: Note that officially the library does not use "slash command" terminology, instead it uses "app commands" as a shortened form of "application command". In the user facing documentation it might be fine since there aren't any hybrid context menu commands so nothing might have to be done here.

This allows you to define a command as both slash and text command without writing separate code for
both counterparts.

> One of the goals of hybrid commands is to also allow you to easily transit from text commands to

Reviewer: Not particularly. The goal is just to allow you to share functionality seamlessly.

sync your :class:`~app_commands.CommandTree` by calling :class:`~app_commands.CommandTree.sync` in order
for slash command to appear.

.. image:: /images/commands/hybrid1.png
.. image:: /images/commands/hybrid2.png

You can create hybrid command groups and sub-commands using the :meth:`.Bot.hybrid_group`
decorator.

.. code-block:: python3

    @bot.hybrid_group(fallback="get")
    async def tag(ctx, name):
        await ctx.send(f"Showing tag: {name}")

    @tag.command()
    async def create(ctx, name):
        await ctx.send(f"Created tag: {name}")

Due to a Discord limitation, Slash command groups cannot be invoked directly so the ``fallback``
parameter allows you to create a sub-command that will be bound to callback of parent group.

.. image:: /images/commands/hybrid3.png
.. image:: /images/commands/hybrid4.png

Due to certain limitation on slash commands, Certain features of text commands are not supported
on hybrid commands. You can define a hybrid command as long as it meets the same subset that is
supported for slash commands.

Following are **not supported** by hybrid commands:

> - Variadic arguments

Reviewer: Probably elaborate.

> - Nested command groups more than one

Reviewer: ```suggestion
- Group commands with a depth greater than 1.
```

> - Many :class:`~typing.Union` types

Reviewer: ```suggestion
- Most :class:`typing.Union` types.
    - Unions of channel types are allowed
    - Unions of user types are allowed
    - Unions of user types with roles are allowed
```

> Apart from that, All other features such as converters, checks, autocomplete, flags etc.

Reviewer: Should mention that the decorator order matters.

    Be careful on how you write your global checks, as it could also lock you out of your own bot.

.. need a note on global check once here I think

Hybrid Commands
---------------

.. versionadded:: 2.0

:class:`.commands.HybridCommand` is a command that can be invoked as both a text and a slash command.
This allows you to define a command as both slash and text command without writing separate code for
both counterparts.

> The goal of hybrid commands is to allow you to seamlessly share functionality of both command types.

Reviewer: ```suggestion
```

In order to define a hybrid command, The command callback should be decorated with
:meth:`.Bot.hybrid_command` decorator.

.. code-block:: python3

    @bot.hybrid_command()
    async def test(ctx):
        await ctx.send("This is a hybrid command!")

The above command can be invoked as both text and slash command. Note that you have to manually
sync your :class:`~app_commands.CommandTree` by calling :class:`~app_commands.CommandTree.sync` in order

> for slash command to appear.

Reviewer: ```suggestion
for slash commands to appear.
```

.. image:: /images/commands/hybrid1.png
.. image:: /images/commands/hybrid2.png

You can create hybrid command groups and sub-commands using the :meth:`.Bot.hybrid_group`
decorator.

.. code-block:: python3

    @bot.hybrid_group(fallback="get")
    async def tag(ctx, name):
        await ctx.send(f"Showing tag: {name}")

    @tag.command()
    async def create(ctx, name):
        await ctx.send(f"Created tag: {name}")

> Due to a Discord limitation, Slash command groups cannot be invoked directly so the ``fallback``

Reviewer: ```suggestion
Due to a Discord limitation, slash command groups cannot be invoked directly so the ``fallback``
```

parameter allows you to create a sub-command that will be bound to callback of parent group.

.. image:: /images/commands/hybrid3.png
.. image:: /images/commands/hybrid4.png

> Due to certain limitation on slash commands, Certain features of text commands are not supported

Reviewer: ```suggestion
Due to certain limitations on slash commands, some features of text commands are not supported
```

on hybrid commands. You can define a hybrid command as long as it meets the same subset that is
supported for slash commands.

> Following are **not supported** by hybrid commands:

Reviewer: ```suggestion
Following are currently **not supported** by hybrid commands:
```

- Variable number of arguments. e.g. ``*arg: int``
- Group commands with a depth greater than 1.
- Most :class:`typing.Union` types.
    - Unions of channel types are allowed
    - Unions of user types are allowed
    - Unions of user types with roles are allowed

> Apart from that, All other features such as converters, checks, autocomplete, flags etc.

Reviewer: ```suggestion
Apart from that, all other features such as converters, checks, autocomplete, flags etc.
```

> are supported on hybrid commands. Note that for decorators related to application commands

Reviewer: ```suggestion
are supported on hybrid commands. Note that due to a design constraint, decorators related to application commands
```

> such as :func:`discord.app_commands.autocomplete` etc. should be placed below the :meth:`.Bot.hybrid_command`

Reviewer: ```suggestion
such as :func:`discord.app_commands.autocomplete` should be placed below the :func:`~ext.commands.hybrid_command` decorator.
```

decorator.

> For convenience and ease in writing code, The :class:`~ext.commands.Context` class implements

Reviewer: Probably mention that `Context.interaction` is now a thing too.
